qiskit_ode/signals/signals.py

```diff
@@ -27,167 +28,196 @@
+            carrier_freq: Frequency of the carrier.
+            phase: The phase of the carrier.
+            name: Name of the signal.
+        """
         self._name = name
-        self._carrier_freq = Array(0.0)
-        self._phase = Array(0.0)
 
-    @property
-    def carrier_freq(self):
-        """Return the carrier frequency of the signal."""
-        return self._carrier_freq
+        if isinstance(envelope, (float, int)):
+            envelope = Array(complex(envelope))
 
-    @property
-    def phase(self):
-        """Return the phase of the Signal."""
-        return self._phase
+        if isinstance(envelope, Array):
+            self._envelope = lambda t: envelope
 
-    @abstractmethod
-    def conjugate(self):
-        """Return a new signal that is the complex conjugate of self."""
+        if callable(envelope):
+            self._envelope = envelope
 
-    @abstractmethod
-    def envelope_value(self, t: float = 0.0) -> complex:
-        """Evaluates the envelope at time t."""
+        # initialize internally stored carrier/phase information
+        self._carrier_freq = None
+        self._phase = None
+        self._carrier_arg = None
+        self._phase_arg = None
 
-    @abstractmethod
-    def value(self, t: float = 0.0) -> float:
-        """Return the value of the signal at time t."""
+        # set carrier and phase
+        self.carrier_freq = carrier_freq
+        self.phase = phase
 
     @property
     def name(self) -> str:
         """Return the name of the signal."""
         return self._name
 
-    def to_pwc(self, dt: float, n_samples: int, start_time: float = 0.0) -> "PiecewiseConstant":
-        """
-        Converts a signal to a piecewise constant signal.
+    @property
+    def carrier_freq(self) -> Array:
+        """The carrier frequency of the signal."""
+        return self._carrier_freq
 
-        Args:
-            dt: Time increment to use.
-            n_samples: number of steps to resample with.
-            start_time: start time from which to resample.
+    @carrier_freq.setter
+    def carrier_freq(self, carrier_freq: float):
+        """Carrier frequency setter."""
+        if type(carrier_freq) == list:
+            carrier_freq = [Array(entry).data for entry in carrier_freq]
+        self._carrier_freq = Array(carrier_freq)
+        self._carrier_arg = 1j * 2 * np.pi * self._carrier_freq
 
-        Returns:
-            A piecewiseConstant signal.
-        """
-        samples = [self.envelope_value(dt * idx + start_time) for idx in range(n_samples)]
+    @property
+    def phase(self) -> Array:
+        """The phase of the signal."""
+        return self._phase
 
-        return PiecewiseConstant(
-            dt, samples, start_time=start_time, carrier_freq=self.carrier_freq, phase=self.phase
-        )
+    @phase.setter
+    def phase(self, phase: float):
+        """Phase setter."""
+        if type(phase) == list:
+            phase = [Array(entry).data for entry in phase]
```

Comment:
Same comment as above.

Reply:
Same reply: moved the list handling version to `SignalSum`.